sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/AuthenticationProvider.java

```diff
@@ -0,0 +1,199 @@
+        mTokenProvider = tokenProvider;
+    }
+
+    /**
+     * Get the type of this authentication provider.
+     *
+     * @return the type.
+     */
+    Type getType() {
+        return mType;
+    }
+
+    /**
+     * Get the ticket key for this authentication provider.
+     *
+     * @return the ticket key.
+     */
+    String getTicketKey() {
+        return mTicketKey;
+    }
+
+    /**
+     * Get the ticket key hash for this authentication provider.
+     *
+     * @return the ticket key hash.
+     */
+    String getTicketKeyHash() {
+        return mTicketKeyHash;
+    }
+
+    /**
+     * Token provider object that will be called to provide a current authentication token.
+     *
+     * @return the token provider.
+     */
+    TokenProvider getTokenProvider() {
+        return mTokenProvider;
+    }
+
+    /**
+     * Call token provider callback in background.
+     */
+    void acquireTokenAsync() {
+        AppCenterLog.debug(LOG_TAG, "Calling token provider=" + mType + " callback.");
+        mTokenProvider.getToken(mTicketKey, new AuthenticationCallback() {
+
+            @Override
+            public void onAuthenticationResult(String token, Date expiresAt) {
+                handleTokenUpdate(token, expiresAt);
+            }
+        });
+    }
+
+    /**
+     * Handle token callback update.
+     *
+     * @param token     token value.
+     * @param expiresAt token expiry date.
+     */
+    private synchronized void handleTokenUpdate(String token, Date expiresAt) {
+
+        /* Check parameters. */
+        AppCenterLog.debug(LOG_TAG, "Got result back from token provider=" + mType);
+        if (token == null) {
+            AppCenterLog.error(LOG_TAG, "Authentication failed for ticketKey=" + mTicketKey);
+            return;
+        }
+        if (expiresAt == null) {
+            AppCenterLog.error(LOG_TAG, "No expiry provided for ticketKey=" + mTicketKey);
+            return;
+        }
+
+        /* Update cache. */
+        TicketCache.putTicket(mTicketKeyHash, token);
+
+        /* Schedule refresh. */
+        long refreshTime = (long) ((expiresAt.getTime() - System.currentTimeMillis()) * REFRESH_THRESHOLD);
+        AppCenterLog.info(LOG_TAG, "User authenticated for " + refreshTime + " ms. for provider=" + mType);
+        mRefreshTimer = new Runnable() {
+
+            @Override
+            public void run() {
+                acquireTokenAsync();
+            }
+        };
+        HandlerUtils.getMainHandler().postDelayed(mRefreshTimer, refreshTime);
+    }
+
+    /**
+     * Cancel token refresh timer if any is pending.
+     */
+    synchronized void stopRefreshing() {
+        if (mRefreshTimer != null) {
+            HandlerUtils.getMainHandler().removeCallbacks(mRefreshTimer);
+            mRefreshTimer = null;
+        }
+    }
+
+    /**
+     * The supported types of the authentication provider.
+     */
+    public enum Type {
+
+        /**
+         * Microsoft authentication.
```

Comment:
nit - "Microsoft account authentication"

Technically "MSA" stands for "Microsoft Account", see https://en.wikipedia.org/wiki/Microsoft_account.